server_test.go

```diff
@@ -569,3 +584,262 @@ func assertCSVFieldsMatch(t *testing.T, expected, actual [][]string, columns []i
+		serverKey              string
+		serverCertificate      string
+		authorityCertificate   string
+		expectedConnectResults [4]bool
+	}{
+		{"TCP", "", "", "", [4]bool{true, false, false, false}},
+		{"encrypted", serverKey, serverCertificate, "", [4]bool{false, true, true, true}},
+		{"authenticated", serverKey, serverCertificate, caCertificate, [4]bool{false, false, false, true}},
+	}
+
+	// load all the various keys and certificates for the client
+	trustServerCA := x509.NewCertPool()
+	ok := trustServerCA.AppendCertsFromPEM([]byte(caCertificate))
+	if !ok {
+		t.Fatal("could not load server certificate")
+	}
+	wrongCert, err := tls.X509KeyPair([]byte(wrongClientCertificate), []byte(clientKey))
+	if err != nil {
+		t.Fatal("could not load wrong client cert/key:", err)
+	}
+	wrongConfig := &tls.Config{
+		RootCAs:      trustServerCA,
+		Certificates: []tls.Certificate{wrongCert},
+	}
+	correctCert, err := tls.X509KeyPair([]byte(correctClientCertificate), []byte(clientKey))
+	if err != nil {
+		t.Fatal("could not load correct client cert/key:", err)
+	}
+	correctConfig := &tls.Config{
+		RootCAs:      trustServerCA,
+		Certificates: []tls.Certificate{correctCert},
+	}
+
+	// all supported client configurations
+	clientConfigs := []struct {
+		name      string
+		tlsConfig *tls.Config
+	}{
+		{"TCP", nil},
+		{"TLS no cert", &tls.Config{RootCAs: trustServerCA}},
+		{"TLS wrong cert", wrongConfig},
+		{"TLS correct cert", correctConfig},
+	}
+
+	for _, serverConfig := range serverConfigs {
+		config := localConfig()
+		config.NumWorkers = 1
+		// Use a unique port to avoid race with shutting down accept goroutine on Linux
+		config.TcpAddress = fmt.Sprintf("localhost:%d", HTTPAddrPort)
+		HTTPAddrPort++
+		config.TLSKey = serverConfig.serverKey
+		config.TLSCertificate = serverConfig.serverCertificate
+		config.TLSAuthorityCertificate = serverConfig.authorityCertificate
+		f := newFixture(t, config)
+		defer f.Close() // ensure shutdown if the test aborts
+
+		// attempt to connect and send stats with each of the client configurations
+		for i, clientConfig := range clientConfigs {
+			expectedSuccess := serverConfig.expectedConnectResults[i]
+			err := sendTCPMetrics(config.TcpAddress, clientConfig.tlsConfig, f)
+			if err != nil {
+				if expectedSuccess {
+					t.Errorf("server config: '%s' client config: '%s' failed: %s",
+						serverConfig.name, clientConfig.name, err.Error())
+				} else {
+					fmt.Printf("SUCCESS server config: '%s' client config: '%s' got expected error: %s\n",
+						serverConfig.name, clientConfig.name, err.Error())
+				}
+			} else if !expectedSuccess {
+				t.Errorf("server config: '%s' client config: '%s' worked; should fail!",
+					serverConfig.name, clientConfig.name)
+			} else {
+				fmt.Printf("SUCCESS server config: '%s' client config: '%s'\n",
+					serverConfig.name, clientConfig.name)
+			}
+		}
+
+		f.Close()
+	}
+}
+
+// DO NOT USE: insecure test keys generated with the following commands:
```

Comment:
Can we either put these in a separate `.go` file (rather than `server_test.go`) or in a fixture (like we do for the JSON examples? My preference is for the latter, since it's more consistent, but either one works. 

If we use a fixture, we could read the certs in the `init()` for the tests and include this comment there.